### PR TITLE
TNT-41046 Encapsulate telemetry reporting to make it easy to switch storage layer

### PR DESCRIPTION
**Pull request by andyruwruw**
## Description

Telemetry related logic was moved out of the target-decisioning-engine notificationProvider and into target-tools telemetryProvider.

The TelemetryProvider stores telemetry entries, and can add them to a DeliveryRequest.

Currently the implementation was kept relatively the same. The NotificationProvider has a TelemetryProvider, using it to store entries, and to add telemetries before sending notifications.

## Related Issue

[Jira Ticket](https://jira.corp.adobe.com/browse/TNT-41046)

## Motivation and Context

Description from Jira Ticket:

Currently telemetry reporting is done through Delivery API and makes its way into Prometheus; reporting and storage are tightly coupled in the SDKs.  We should decouple them and create an abstract storage service in case we switch to a different storage layer in the future.

## How Has This Been Tested?

Tests were added for the new Telemetry Provider, just ensuring it stores entries and adds them to a request.
Tests were run with `npm run test` and written using jest.

This implementation will impact how the notification provider stores telemetry entries by encapsulating telemetry related logic in a separate class.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html). I'm an intern so I need to contact someone about resolving this.
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: packages/target-decisioning-engine/src/notificationProvider.js [hunk 1]
```diff
@@ -26,11 +26,14 @@ function NotificationProvider(
   sendNotificationFunc = noop,
   telemetryEnabled = true
 ) {
-  const { requestId } = request;
   const timestamp = now();
   const prevEventKeys = new Set();
   let notifications = [];
-  let telemetryEntries = [];
+  const telemetryProvider = TelemetryProvider(
```
Comment (andyruwruw) [inline packages/target-decisioning-engine/src/notificationProvider.js:32]: Change order, leave off final param

File: packages/target-tools/src/index.js [hunk 2]
```diff
@@ -81,6 +81,7 @@ export { getFetchWithRetry, getFetchApi } from "./networking";
 
 export { AttributesProvider } from "./attributesProvider";
 export { EventProvider } from "./eventProvider";
+export { default as TelemetryProvider } from "./telemetryProvider";
```
Comment (andyruwruw) [inline packages/target-tools/src/index.js:84]: Change to export const

File: packages/target-tools/src/telemetryProvider.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import { now } from "./lodash";
+import { DECISIONING_METHOD } from "./enums";
+
+/**
+ * The get TelemetryProvider initialization method
+ * @param {function} sendTelemetriesFunc function used to send the telemetries, required
+ */
+function TelemetryProvider(
```
Comment (andyruwruw) [inline packages/target-tools/src/telemetryProvider.js:8]: export function

File: packages/target-tools/src/telemetryProvider.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        decisioningMethod
+      },
+      ...entry
+    });
+  }
+
+  function sendTelemetries(deliveryRequest) {
+    if (telemetryEntries.length > 0) {
+      const deliveryRequestWithTelemetry = {
+        ...deliveryRequest,
```
Comment (andyruwruw) [inline packages/target-tools/src/telemetryProvider.js:38]: move as last item in object

File: packages/target-tools/src/telemetryProvider.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        telemetry: {
+          entries: telemetryEntries
+        }
+      };
+      telemetryEntries = [];
```
Comment (andyruwruw) [inline packages/target-tools/src/telemetryProvider.js:43]: reset() function instead

File: packages/target-tools/src/telemetryProvider.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      requestId,
+      timestamp,
+      features: {
+        decisioningMethod
+      },
+      ...entry
+    });
+  }
+
+  function sendTelemetries(deliveryRequest) {
```
Comment (andyruwruw) [inline packages/target-tools/src/telemetryProvider.js:35]: pass in function as class param

Comment (andyruwruw) [inline packages/target-tools/src/telemetryProvider.js:35]: `executeTelemetries`

File: packages/target-tools/src/telemetryProvider.spec.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    const mockExecute = jest.fn();
+
+    const provider = TelemetryProvider(TARGET_REQUEST, mockExecute);
+
+    provider.addEntry(TARGET_TELEMETRY_ENTRY);
+    provider.executeTelemetries(TARGET_REQUEST);
+
+    expect(mockExecute.mock.calls.length).toBe(1);
+    expect(mockExecute.mock.calls[0][1][0]).toEqual(
+      expect.objectContaining({
```
Comment (dcottingham) [inline packages/target-tools/src/telemetryProvider.spec.js:23]: Let's add the `execution: 1` key/value to this assertion

File: packages/target-tools/src/telemetryProvider.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import { noop } from "./utils";
+import { DECISIONING_METHOD } from "./enums";
+
+/**
+ * The get TelemetryProvider initialization method
+ * @param {function} sendTelemetriesFunc function used to send the telemetries, required
+ */
+export function TelemetryProvider(
+  request,
+  executeTelemetriesFunc = noop,
```
Comment (dcottingham) [inline packages/target-tools/src/telemetryProvider.js:13]: I think we should remove `noop` as the default value for `executeTelemetriesFunc`, and just make it a required arg w/ no default

File: packages/target-tools/src/telemetryProvider.spec.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+          decisioningMethod: expect.any(String)
+        }
+      })
+    );
+
+    const entries = provider.getEntries();
+    expect(entries.length).toEqual(0);
+  });
+
+  it("execute function undefined", () => {
```
Comment (dcottingham) [inline packages/target-tools/src/telemetryProvider.spec.js:36]: I guess we don't need this after all if we get rid of the noop default

**Review by dcottingham — COMMENTED** [review]
Looks great!  Just a couple minor comments

**Comment by andyruwruw** [comment]
Added the edits!

File: packages/target-tools/src/telemetryProvider.spec.js [hunk 10]
```diff
@@ -28,22 +27,12 @@ describe("TelemetryProvider", () => {
         }
       })
     );
+    expect(mockExecute.mock.calls[0][1][0].execution).toBe(1);
```
Comment (dcottingham) [inline packages/target-tools/src/telemetryProvider.spec.js:30]: Minor nitpick for sake of consistency - let's remove this line and add it into the previous assertion:
`expect.objectContaining({........,
    execution: 1
})`

**Pull request merged**